### PR TITLE
cdao backend canister changes

### PR DESCRIPTION
**Pull request by debjit-bhowal-rs**
#### To check the status of the deployment

- check if the platform orchestrator performed the step to upgrade subnet canister with appropriate version: [Platform Orchestrator function](https://dashboard.internetcomputer.org/canister/74zq4-iqaaa-aaaam-ab53a-cai#get_subnet_last_upgrade_status)
- check the status of upgrade for individual canisters in subnet orchestrators and verify the version. Example for one of the subnet orchesrator: [Subnet Orchestrator function](https://dashboard.internetcomputer.org/canister/rimrc-piaaa-aaaao-aaljq-cai#get_index_details_last_upgrade_status)

To test the platform orchestrator, you can use the following command:

`dfx canister call --query 74zq4-iqaaa-aaaam-ab53a-cai get_subnet_last_upgrade_status --network=ic`
 the canister id (74zq4-iqaaa-aaaam-ab53a-cai) is taken from the canister_ids.json file

### CODE REVIEW

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            controllers: Some(controllers),
+            ..Default::default()
+        },
+    })
+    .await?;
+    Ok(())
+}
+
+#[query]
+async fn deployed_cdao_canisters() -> Vec<DeployedCdaoCanisters> {
```
Comment (ravibazz) [inline src/canister/individual_user_template/src/api/cdao/mod.rs:80]: Should we use better naming here?

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    CANISTER_DATA.with(|cdata| {
+        cdata
+            .borrow()
+            .cdao_canisters
+            .clone()
+    })
+}
+
+#[update]
```
Comment (ravibazz) [inline src/canister/individual_user_template/src/api/cdao/mod.rs:89]: Do we need to add guard for this call?

**Review by ravibazz — CHANGES_REQUESTED** [review]

**Review by rupansh — CHANGES_REQUESTED** [review]
Please also add `CanisterIdSnsWasm` to known_principal_ids in a post_upgrade step

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+                .copied()
+        })
+        .expect("Super admin not specified in config");
+
+    let user_can = ic_cdk::id();
+    let mut update_ctrls_futs = FuturesUnordered::new();
+    update_ctrls_futs.push(update_controllers(
+        governance,
+        vec![admin_canister, user_can, root.0],
+    ));
```
[the remark this answers is not in the document]
Comment (rupansh) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:237]: asked by @saikatdas0790

File: src/canister/individual_user_template/src/api/cdao/token.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    CANISTER_DATA.with(|cdata| {
+        let mut cdata = cdata.borrow_mut();
+        cdata.token_roots.insert(root_canister, ());
+    });
+
+
+    return Ok(true);
+}
+
+#[update]
```
Comment (rupansh) [inline src/canister/individual_user_template/src/api/cdao/token.rs:40]: We need to add a guard here so only the canister owner can call

File: src/canister/individual_user_template/src/data_model/mod.rs [hunk 5]
```diff
@@ -85,6 +86,11 @@ pub struct CanisterData {
     pub device_identities: Vec<DeviceIdentity>,
     #[serde(default)]
     pub ml_feed_cache: Vec<MLFeedCacheItem>,
+    #[serde(default)]
+    pub cdao_canisters: Vec<DeployedCdaoCanisters>,
+    // list of root token canisters 
+    #[serde(skip, default = "_default_token_list")]
+    pub token_roots: ic_stable_structures::btreemap::BTreeMap<Principal, (), Memory>,
```
[the remark this answers is not in the document]
Comment (rupansh) [reply src/canister/individual_user_template/src/data_model/mod.rs:93]: `add_token` becomes O(log(n)) as opposed to O(n) https://github.com/go-bazzinga/hot-or-not-backend-canister/pull/388/files/d3329c2b86bd9ec2d967a071a48639cc8d95e815#diff-8f91a3be1e3967c93f44e4aacf37744428f1f1e8fa1aaeda792dd162a08d5ec3R12

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        })
+        .expect("SNS WASM not specified in config");
+
+    // let sns_versions = ic_cdk::call::<_, (HashMap<String, String>,)>(
+    //     sns_wasm,
+    //     "get_latest_sns_version_pretty",
+    //     (),
+    // )
+    // .await?
+    // .0;
```
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:155]: yep, mb

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    canister_specific::individual_user_template::{consts::CDAO_TOKEN_LIMIT, types::{
+        cdao::DeployedCdaoCanisters, error::CdaoDeployError, session::SessionType,
+    }},
+    common::types::known_principal::KnownPrincipalType,
+};
+
+use crate::CANISTER_DATA;
+// 0.1T
+const CDAO_CYCLE_CNT: u128 = 100000000000;
+
```
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:32]: yep that's true, currently 0.1T manages to deploy only 1 token, with an `insufficient cycles` error from subsequent times
better to 5x this value

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 1 above)
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:80]: `tokens_minted`?

File: src/canister/individual_user_template/src/api/cdao/token.rs (hunk 4 above)
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/token.rs:40]: adding

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+    sns_versions.insert("Swap".to_string(), "0098ee44db411c37da5e8281d31df5be9e385c67d6df5fc8004577dc92ed4b17".to_string());
+    sns_versions.insert("Ledger".to_string(), "741745514d330168515cd1752100e4159b7122d5bef01cb1d161321e71c0137b".to_string());
+    sns_versions.insert("Ledger Archive".to_string(), "e4a66f4985dac7b59d1df08e451b71d66b99091899daa25b0274b6208cfccaed".to_string());
+
+    let get_wasm_hash = |name: &str| hex::decode(&sns_versions[name]).unwrap();
+    let gov_hash = get_wasm_hash("Governance");
+    let ledger_hash = get_wasm_hash("Ledger");
+    let root_hash = get_wasm_hash("Root");
+    let swap_hash = get_wasm_hash("Swap");
+    let index_hash = get_wasm_hash("Ledger Index");
```
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:170]: i think we should retain this, so in future to add a function which can modify the `sns_versions` hashmap

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    // .await?
+    // .0;
+
+    let mut sns_versions: HashMap<String, String> = HashMap::new();
+    sns_versions.insert("Governance".to_string(), "a46ca12aecbd6a6c962ab12052cea0ae94f0dcaa65ae0d407d1d72c9121b1074".to_string());
+    sns_versions.insert("Ledger Index".to_string(), "14c48700ba8f68a8af2dc68aa198d3c4f41d4fb02d9439cc81052511442236af".to_string());
+    sns_versions.insert("Root".to_string(), "03326e187fca708aa14a3100013c2be4c94d5f89b439e71ec4c3c0c62e0537d4".to_string());
+    sns_versions.insert("Swap".to_string(), "0098ee44db411c37da5e8281d31df5be9e385c67d6df5fc8004577dc92ed4b17".to_string());
+    sns_versions.insert("Ledger".to_string(), "741745514d330168515cd1752100e4159b7122d5bef01cb1d161321e71c0137b".to_string());
+    sns_versions.insert("Ledger Archive".to_string(), "e4a66f4985dac7b59d1df08e451b71d66b99091899daa25b0274b6208cfccaed".to_string());
```
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs]: fixed

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 2 above)
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:89]: added

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rupansh) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:170]: There's no need to future proof this here, as we're uncertain what the upgrade logic might look like
Wasted cycles is wasted money :P

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (rupansh) [reply src/canister/individual_user_template/src/api/cdao/mod.rs]: @debjit-bhowal-rs are we sure deploy_cdao will work fine with the hashes? from what I can see these were picked up from the latest sns, while our code is not up-to-date with the latest sns

File: src/canister/individual_user_template/src/data_model/mod.rs [hunk 10]
```diff
@@ -118,6 +124,10 @@ pub fn _default_success_history(
     ic_stable_structures::btreemap::BTreeMap::init(get_success_history_memory())
 }
 
+pub fn _default_token_list() -> ic_stable_structures::btreemap::BTreeMap<Principal, (), Memory> {
+    ic_stable_structures::btreemap::BTreeMap::new(get_token_list_memory())
```
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/data_model/mod.rs:128]: done

File: src/canister/individual_user_template/src/data_model/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/data_model/mod.rs:93]: resolving this

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs]: made this change

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs]: i tried the lates commit locally, it works, so the wasm hashes are good to go
@rupansh @rupansh-gob

File: src/canister/individual_user_template/src/api/cdao/mod.rs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        swap,
+        index,
+    };
+    let mut payloads = init_payload
+        .build_canister_payloads(&sns_canisters, None, true)
+        .map_err(CdaoDeployError::InvalidInitPayload)?;
+    let time_seconds = ic_cdk::api::time() / 1_000_000_000;
+    payloads.swap.swap_start_timestamp_seconds = Some(time_seconds);
+    payloads.swap.swap_due_timestamp_seconds = Some(time_seconds + 300);
+    payloads.swap.icp_ledger_canister_id = "ryjl3-tyaaa-aaaaa-aaaba-cai".to_string();
```
Comment (rupansh-gob) [inline src/canister/individual_user_template/src/api/cdao/mod.rs:144]: is it fine to set it to this?

**Review by rupansh-gob — CHANGES_REQUESTED** [review]
small doubt

File: src/canister/individual_user_template/src/api/cdao/mod.rs (hunk 11 above)
Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:144]: - `ryjl3-tyaaa-aaaaa-aaaba-cai` this is anyways the mainnet canister, we are setting this just so local testing is easier (for eg there's a setup_icp_ledger.sh to deploy the icp ledger on local)
- the timeout of 5 mins is the initial swap time. 5 min is the max, can also set lower. i was considering if i should make it a separate input param itself to `deploy_cado_sns`

Comment (debjit-bhowal-rs) [reply src/canister/individual_user_template/src/api/cdao/mod.rs:145]: i added swap_time as a parameter directly

**Review by rupansh-gob — APPROVED** [review]

**Review by rupansh-gob — DISMISSED** [review]

File: src/lib/integration_tests/tests/creator_dao/main.rs [hunk 12]
```diff
@@ -44,7 +45,7 @@ use test_utils::setup::{
 };
 use ic_base_types::PrincipalId;
 use ic_sns_wasm::init::SnsWasmCanisterInitPayload;
-pub type CanisterId = Principal;
+// pub type CanisterId = Principal;
```
Comment (ravibazz) [inline src/lib/integration_tests/tests/creator_dao/main.rs:48]: Lets remove the comments

File: src/lib/integration_tests/tests/creator_dao/main.rs [hunk 13]
```diff
@@ -492,4 +504,217 @@ fn creator_dao_tests() {
...
+
+    let mut swap_canister: Option<Principal> = None;
+    let mut gov_canister: Option<Principal> = None;
+    let mut ledger_canister: Option<Principal> = None;
+    if res.len() > 0 {
+        let res = res[0].clone();
+        swap_canister = Some(res.swap);
+        gov_canister = Some(res.governance);
+        ledger_canister = Some(res.ledger);
+    }
```
Comment (ravibazz) [inline src/lib/integration_tests/tests/creator_dao/main.rs:523]: add assert for `res.len` and directly assign canister principal to variables.

File: src/lib/integration_tests/tests/creator_dao/main.rs [hunk 14]
```diff
@@ -492,4 +504,217 @@ fn creator_dao_tests() {
...
+    ).map(|res| {
+        let response = match res {
+            WasmResult::Reply(payload) => Decode!(&payload, Nat).unwrap(),
+            _ => panic!("\n🛑 get requester principals canister id failed\n"),
+        };
+        response
+    }).unwrap();
+    ic_cdk::println!("🧪 Result: {:?}", res);
+
+    if let Some(swap_canister) = swap_canister {
```
Comment (ravibazz) [inline src/lib/integration_tests/tests/creator_dao/main.rs:555]: remove if statements use assert and unwrap instead.

File: src/lib/integration_tests/tests/creator_dao/main.rs [hunk 15]
```diff
@@ -492,4 +504,217 @@ fn creator_dao_tests() {
...
+        ic_cdk::println!("🧪 SNS token Balance of alice: {:?}", res);
+
+        let expected_balance = Nat::from(4_400_000 - tx_fee);
+        ic_cdk::println!("🧪 Expected Balance: {:?}", expected_balance);
+
+        assert!(res == expected_balance);
+
+    } else {
+        panic!("Ledger Canister not found, something is wrong");
+    }
```
Comment (ravibazz) [inline src/lib/integration_tests/tests/creator_dao/main.rs:719]: We can remove the panic statements as well once we remove if statements

**Review by ravibazz — CHANGES_REQUESTED** [review]
Minor changes. Functionality wise looks good. Will approve once the comments are addressed. Overall good work.

File: dfx.json [hunk 16]
```diff
@@ -1,5 +1,25 @@
...
+      "remote": {
+        "id": {
+          "ic": "ryjl3-tyaaa-aaaaa-aaaba-cai"
+        }
+      }
+    },
+    "icp_index": {
+      "type": "custom",
+      "candid": "candid/index.did",
+      "wasm": "./icp_index.wasm.gz",
```
Comment (ravibazz) [inline dfx.json:16]: update these settings `wasm` is pointing to wrong files

**Review by ravibazz — CHANGES_REQUESTED** [review]
Verify if the local deployment is working. `dfx.json` is looks incorrect

File: scripts/canisters/local_deploy/install_all_canisters.sh [hunk 17]
```diff
@@ -24,6 +24,9 @@ dfx canister create --no-wallet post_cache
 dfx canister create --no-wallet user_index
 dfx canister create --no-wallet platform_orchestrator
 
+scripts/canisters/local_deploy/create_sns_wasm.sh
+
+dfx build individual_user_template
```
Comment (ravibazz) [inline scripts/canisters/local_deploy/install_all_canisters.sh:29]: don't need these now.

**Review by ravibazz — CHANGES_REQUESTED** [review]

File: dfx.json (hunk 16 above)
Comment (ravibazz) [reply dfx.json:16]: Ahh I think the icp_index and ledger looks good. Don't worry about this. My bad.

**Review by ravibazz — APPROVED** [review]

**Pull request merged**